File: .github/workflows/script/jdwp_tester.py
#!/usr/bin/env python3
"""
JDWP Tester - Standalone script for testing JDWP debug connections

This script continuously attempts JDWP handshakes and resume commands
until successful or timeout. It's used to verify that debug port forwarding
is working and can resume suspended JVMs.
"""

import socket
import struct
import sys
import time
import argparse
from pathlib import Path


def recv_exactly(sock: socket.socket, num_bytes: int) -> bytes:
    """Read exactly num_bytes from sock, looping until all bytes are received."""
    data = b""
    while len(data) < num_bytes:
        chunk = sock.recv(num_bytes - len(data))
        if not chunk:
            raise ConnectionError(f"Connection closed after {len(data)}/{num_bytes} bytes")
        data += chunk
    return data


def jdwp_handshake_and_resume(host: str, port: int) -> bool:
    """
    Perform JDWP handshake and send VirtualMachine.Resume command.

    Returns True if successful, False otherwise.
    """
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)

        # Connect and handshake
        sock.connect((host, port))
        sock.sendall(b"JDWP-Handshake")
        response = recv_exactly(sock, 14)

        if response != b"JDWP-Handshake":
            print(f"[JDWP] Handshake failed, got: {response!r}", file=sys.stderr)
            return False

        # Send VirtualMachine.Resume command (CommandSet=1, Command=9)
        cmd_set, cmd = 1, 9
        length = struct.pack(">I", 11)
        packet_id = struct.pack(">I", 1)
        flags = b"\x00"
        packet = length + packet_id + flags + bytes([cmd_set, cmd])

        sock.sendall(packet)
        recv_exactly(sock, 11)

        sock.close()
        print("[JDWP] Handshake + Resume successful")
        return True

    except Exception as e:
        print(f"[JDWP] Connection failed: {e}", file=sys.stderr)
        return False


def main():
    parser = argparse.ArgumentParser(description="Test JDWP debug connection")
    parser.add_argument("host", help="Debug host (usually localhost)")
    parser.add_argument("port", type=int, help="Debug port (usually 5005)")
    parser.add_argument("--timeout", type=int, default=300, help="Total timeout in seconds")
    parser.add_argument("--stop-file", default="/tmp/solo-jdwp-stop", help="Stop signal file")

    args = parser.parse_args()

    deadline = time.time() + args.timeout
    success_count = 0

    print(f"[JDWP] Starting probe for {args.host}:{args.port} (timeout: {args.timeout}s)")

    while True:
        # Check for stop signal
        if Path(args.stop_file).exists():
            if success_count > 0:
                print(f"[JDWP] Stop requested after {success_count} successful resume(s)")
                return 0
            else:
                print("[JDWP] Stop requested but no successful resume observed", file=sys.stderr)
                return 1

        # Try JDWP handshake and resume
        if jdwp_handshake_and_resume(args.host, args.port):
            success_count += 1
            print(f"[JDWP] Success count: {success_count}")

        # Check timeout
        if time.time() >= deadline:
            if success_count > 0:
                print(f"[JDWP] Timeout reached with {success_count} successful resume(s)")
                return 0
            else:
                print(f"[JDWP] Timed out after {args.timeout}s without any successful handshake/resume", file=sys.stderr)
                return 1

        time.sleep(2)


if __name__ == "__main__":
    sys.exit(main())

File: .github/workflows/script/jdwp-tester.ts
#!/usr/bin/env node

/**
 * JDWP Tester - Standalone script for testing JDWP debug connections
 *
 * This script continuously attempts JDWP handshakes and resume commands
 * until successful or timeout. It's used to verify that debug port forwarding
 * is working and can resume suspended JVMs.
 */

import net, { type Socket } from 'node:net';
import fs from 'node:fs';
import { setTimeout as delay } from 'node:timers/promises';
import process from 'node:process';

function recvExactly(socket: Socket, numBytes: number): Promise<Buffer> {
	return new Promise((resolve, reject): void => {
		const data: Buffer[] = [];
		let totalBytes = 0;

		const onData = (chunk: Buffer): void => {
			data.push(chunk);
			totalBytes += chunk.length;

			if (totalBytes >= numBytes) {
				socket.removeListener('data', onData);
				socket.removeListener('error', onError);
				socket.removeListener('close', onClose);
				resolve(Buffer.concat(data, numBytes));
			}
		};

		const onError = (error: Error): void => {
			socket.removeListener('data', onData);
			socket.removeListener('close', onClose);
			reject(error);
		};

		const onClose = (): void => {
			socket.removeListener('data', onData);
			socket.removeListener('error', onError);
			reject(
				new Error(
					`Connection closed after ${totalBytes}/${numBytes} bytes`
				)
			);
		};

		socket.on('data', onData);
		socket.on('error', onError);
		socket.on('close', onClose);
	});
}

async function jdwpHandshakeAndResume(
	host: string,
	port: number
): Promise<boolean> {
	const socket = new net.Socket();

	try {
		socket.setTimeout(5000);

		// Connect and handshake
		await new Promise<void>((resolve, reject): void => {
			socket.connect(port, host, (): void => {
				resolve();
			});
			socket.on('error', (error: Error): void => {
				reject(error);
			});
		});

		socket.write('JDWP-Handshake');
		const response = await recvExactly(socket, 14);

		if (!response.equals(Buffer.from('JDWP-Handshake'))) {
			console.error(
				`[JDWP] Handshake failed, got: ${response.toString('hex')}`
			);
			return false;
		}

		// Send VirtualMachine.Resume command (CommandSet=1, Command=9)
		const cmdSet = 1;
		const cmd = 9;
		const length = Buffer.alloc(4);
		length.writeUInt32BE(11, 0);
		const packetId = Buffer.alloc(4);
		packetId.writeUInt32BE(1, 0);
		const flags = Buffer.from([0x00]);
		const packet = Buffer.concat([length, packetId, flags, Buffer.from([cmdSet, cmd])]);

		socket.write(packet);
		await recvExactly(socket, 11);

		socket.end();
		console.log('[JDWP] Handshake + Resume successful');
		return true;
	} catch (error) {
		if (error instanceof Error) {
			console.error(`[JDWP] Connection failed: ${error.message}`);
		} else {
			console.error('[JDWP] Connection failed: Unknown error');
		}
		return false;
	} finally {
		socket.destroy();
	}
}

async function main(): Promise<number> {
	const args = process.argv.slice(2);
	let host = 'localhost';
	let port = 5005;
	let timeout = 300;
	let stopFile = '/tmp/solo-jdwp-stop';

	// Simple argument parsing
	for (let i = 0; i < args.length; i += 1) {
		const arg = args[i];

		if (arg === '--timeout') {
			timeout = Number.parseInt(args[i + 1], 10);
			i += 1;
		} else if (arg === '--stop-file') {
			stopFile = args[i + 1];
			i += 1;
		} else if (!arg.startsWith('--') && i < 2) {
			if (i === 0) {
				host = arg;
			} else if (i === 1) {
				port = Number.parseInt(arg, 10);
			}
		}
	}

	if (!host || !port) {
		console.error('Usage: jdwp-tester <host> <port> [--timeout SECONDS] [--stop-file PATH]');
		return 1;
	}

	const deadline = Date.now() + timeout * 1000;
	let successCount = 0;

	console.log(
		`[JDWP] Starting probe for ${host}:${port} (timeout: ${timeout}s)`
	);

	while (true) {
		// Check for stop signal
		if (fs.existsSync(stopFile)) {
			if (successCount > 0) {
				console.log(
					`[JDWP] Stop requested after ${successCount} successful resume(s)`
				);
				return 0;
			} else {
				console.error('[JDWP] Stop requested but no successful resume observed');
				return 1;
			}
		}

		// Try JDWP handshake and resume
		const success = await jdwpHandshakeAndResume(host, port);
		if (success) {
			successCount += 1;
			console.log(`[JDWP] Success count: ${successCount}`);
		}

		// Check timeout
		if (Date.now() >= deadline) {
			if (successCount > 0) {
				console.log(
					`[JDWP] Timeout reached with ${successCount} successful resume(s)`
				);
				return 0;
			} else {
				console.error(
					`[JDWP] Timed out after ${timeout}s without any successful handshake/resume`
				);
				return 1;
			}
		}

		await delay(2000);
	}
}

process.exit(await main());
